### PR TITLE
Test connectivity

### PR DESCRIPTION
**Pull request by Daniel-GrunbergerCA**
1. Add test registry connectivity command
2.  Change flow to support reading information from command as well (backward compatible)
3. Move part of the functions to registryScan object 
4. Remove ".sig" tags from scans (these are tags created by image signatures, and not valid images for scanning)

### CODE REVIEW

File: mainhandler/vulnscan.go [hunk 1]
```diff
@@ -94,143 +96,124 @@ func convertImagesToWebsocketScanCommand(registry *registryScan, sessionObj *uti
...
 
-func (actionHandler *ActionHandler) getRegistryConfig(registryName string) (*registryScanConfig, string, error) {
-	configMap, err := actionHandler.k8sAPI.GetWorkload(armotypes.KubescapeNamespace, "ConfigMap", registryScanConfigmap)
-	// in case of an error or missing configmap, fallback to the deprecated namespace
-	if err != nil || configMap == nil {
-		configMap, err = actionHandler.k8sAPI.GetWorkload(armotypes.ArmoSystemNamespace, "ConfigMap", registryScanConfigmap)
+func (actionHandler *ActionHandler) testRegistryConnectivity(sessionObj *utils.SessionObj) error {
+	registryScan := NewRegistryScan(actionHandler.k8sAPI)
+
+	err := registryScan.parseRegistryFromCommand(sessionObj)
```
Comment (Bezbran) [inline mainhandler/vulnscan.go:120]: This should be a generic function which loads the necessray fields from CM/secret/request body and return unified object

File: mainhandler/vulnscan.go [hunk 2]
```diff
@@ -94,143 +96,124 @@ func convertImagesToWebsocketScanCommand(registry *registryScan, sessionObj *uti
...
+	if err != nil {
+		glog.Errorf("in testRegistryConnectivity: parseRegistryFromCommand failed with error: ", err.Error())
+		sessionObj.Reporter.SetDetails(fmt.Sprintf("%v failed with err %v", testRegistryInformationStatus, err))
+		return fmt.Errorf("parseRegistryCommand failed with err %v", err)
```
Comment (Bezbran) [inline mainhandler/vulnscan.go:124]: Can't follow the error propagation. We need to send it inside the "errors" field in the job report. the details field is to describe the current phase of the job.

File: mainhandler/vulnscan.go [hunk 3]
```diff
@@ -94,143 +96,124 @@ func convertImagesToWebsocketScanCommand(registry *registryScan, sessionObj *uti
...
-	secret, err := actionHandler.k8sAPI.GetWorkload(armotypes.KubescapeNamespace, "Secret", armotypes.RegistryScanSecretName)
-	if err == nil && secret != nil {
-		return secret, err
+func (actionHandler *ActionHandler) testRegistryConnect(registry *registryScan, sessionObj *utils.SessionObj) error {
+	repos, err := registry.enumerateRepoes()
+	if err != nil {
+		if strings.Contains(strings.ToLower(err.Error()), "unauthorized") || strings.Contains(err.Error(), "DENIED") || strings.Contains(err.Error(), "authentication") {
+			// registry info is good, but authentication failed
+			sessionObj.Reporter.SetDetails(string(testRegistryInformationStatus))
+			sessionObj.Reporter.SendStatus(reporterlib.JobSuccess, true, sessionObj.ErrChan)
```
Comment (Bezbran) [inline mainhandler/vulnscan.go:157]: how come error and job success together?

File: mainhandler/vulnscan.go [hunk 4]
```diff
@@ -94,143 +96,124 @@ func convertImagesToWebsocketScanCommand(registry *registryScan, sessionObj *uti
...
-		return fmt.Errorf("get registry(%s) config failed with err %v", registryName, err)
+		return fmt.Errorf("in 'sendReport' failed to json.Marshal, reason: %v", err)
 	}
+	bodyReader := bytes.NewReader(reqBody)
 
-	glog.Infof("scanRegistries:registry(%s) %s configmap  successful", registryName, configMapMode) // systest depedendent
-	registryScan := NewRegistryScan(registryName, *auth, *conf)
-	return actionHandler.scanRegistry(&registryScan, sessionObj)
-}
+	req, err := http.NewRequest("POST", fmt.Sprintf("%s/k8s/repositoriesToTags?jobID=%v&customerGUID=%v&registryName=%v", actionHandler.repositoriesAndTagsReporter.eventReceiverUrl, params.JobID, params.CustomerGUID, params.RegistryName), bodyReader)
```
Comment (Bezbran) [inline mainhandler/vulnscan.go:209]: use url package to build the URL and query params

**Review by Bezbran — CHANGES_REQUESTED** [review]

File: mainhandler/vulnscan.go (hunk 3 above)
Comment (Daniel-GrunbergerCA) [reply mainhandler/vulnscan.go:158]: information - success
authentication - failed

File: mainhandler/imageregistryhandler.go [hunk 5]
```diff
@@ -88,36 +106,30 @@ func errorWithDocumentationRef(errorMessage string) error {
...
 	falseSkipTLS := false
 	return registryAuth{SkipTLSVerify: &falseSkipTLS, Insecure: &falseInsecure, Kind: kind}
 }
 
-func (rs *registryScan) makeRegistryInterface() (regInterfaces.IRegistry, error) {
-	return regFactory.Factory(&authn.AuthConfig{Username: rs.registryAuth.Username, Password: rs.registryAuth.Password, RegistryToken: rs.registryAuth.RegistryToken}, rs.registry.hostname,
-		regCommon.MakeRegistryOptions(false, *rs.registryAuth.Insecure, *rs.registryAuth.SkipTLSVerify, "", "", rs.registry.projectID, rs.registryAuth.Kind))
-}
-
 func (rs *registryScan) hasAuth() bool {
```
Comment (Bezbran) [inline mainhandler/imageregistryhandler.go:130]: should be IsPrivate

File: mainhandler/imageregistryhandler.go [hunk 6]
```diff
@@ -255,29 +266,24 @@ func (registryScan *registryScan) createTriggerRequestConfigMap(k8sAPI *k8sinter
...
-	if i != -1 {
-		registryScan.registryAuth.Password = registryScan.registryAuth.Password[:i]
-	}
-	repos, err := registryScan.listReposInRegistry()
+	glog.Infof("getImagesForScanning: enumerating repoes...")
+	errChan := make(chan error)
+	repos, err := registryScan.enumerateRepoes()
 	if err != nil {
-		glog.Errorf("listReposInRegistry failed with err %v", err)
+		reporter.SetDetails(fmt.Sprintf("enumerateRepoes failed with err %v", err))
```
Comment (Bezbran) [inline mainhandler/imageregistryhandler.go:273]: Don't send errors in details

File: mainhandler/imageregistryhandler.go [hunk 7]
```diff
@@ -331,7 +341,9 @@ func (registryScan *registryScan) setImageToTagsMap(repo string, reporter datast
 				}
 			}
 		}
-		registryScan.mapImageToTags[registryScan.registry.hostname+"/"+repo] = tags
+		if len(tags) > 0 {
```
Comment (Bezbran) [inline mainhandler/imageregistryhandler.go:344]: Why don't put empty list of tags?
How the user will know they have this repo in the registry if we are ommiting it?

File: mainhandler/imageregistryhandler.go [hunk 8]
```diff
@@ -414,23 +427,42 @@ func (registryScan *registryScan) filterScanLimit(limit int) {
 	registryScan.mapImageToTags = filteredImages
 }
 
+func (registryScan *registryScan) enumerateRepoes() ([]string, error) {
+	// token sometimes includes a lot of dots, so we need to remove them
+	i := strings.Index(registryScan.registryInfo.AuthMethod.Password, ".....")
```
Comment (Bezbran) [inline mainhandler/imageregistryhandler.go:432]: use [TrimLeft](https://pkg.go.dev/strings#TrimLeft)

File: mainhandler/imageregistryhandler.go [hunk 9]
```diff
@@ -414,23 +427,42 @@ func (registryScan *registryScan) filterScanLimit(limit int) {
...
+	var repos, pageRepos []string
+	var nextPage *regCommon.PaginationOption
 
 	firstPage := regCommon.MakePagination(iRegistry.GetMaxPageSize())
 	ctx := context.Background()
 	catalogOpts := regCommon.CatalogOption{IsPublic: !registryScan.hasAuth(), Namespaces: registryScan.registry.projectID}
-	for pageRepos, nextPage, err := iRegistry.Catalog(ctx, firstPage, catalogOpts, regCreds); ; pageRepos, nextPage, err = iRegistry.Catalog(ctx, *nextPage, catalogOpts, regCreds) {
+	for pageRepos, nextPage, err = iRegistry.Catalog(ctx, firstPage, catalogOpts, regCreds); ; pageRepos, nextPage, err = iRegistry.Catalog(ctx, *nextPage, catalogOpts, regCreds) {
 		if err != nil {
 			return nil, err
```
Comment (Bezbran) [inline mainhandler/imageregistryhandler.go:459]: Should we set the job report details to descriptive in what stage we had an error?
Like what is the content of the nextPage or something?

File: mainhandler/imageregistryhandler.go [hunk 10]
```diff
@@ -529,3 +595,159 @@ func (registryScan *registryScan) createTriggerRequestCronJob(k8sAPI *k8sinterfa
...
+func (registryScan *registryScan) setRegistryInfoFromAuth(auth registryAuth, registryInfo *armotypes.RegistryInfo) {
+	registryInfo.AuthMethod.Type = auth.AuthMethod
+	registryInfo.AuthMethod.Username = auth.Username
+	registryInfo.AuthMethod.Password = auth.Password
+	registryInfo.SkipTLSVerify = *auth.SkipTLSVerify
+	registryInfo.IsHTTPs = !*auth.Insecure
+	registryInfo.Kind = string(auth.Kind)
+}
+
+func (registryScan *registryScan) getRegistryConfig(registryInfo *armotypes.RegistryInfo) (string, error) {
```
Comment (Bezbran) [inline mainhandler/imageregistryhandler.go:658]: Is this code will give a precedence to the data from notification uppon the data from CM/secret

Comment (Bezbran) [reply mainhandler/imageregistryhandler.go:658]: ?

File: mainhandler/vulnscan.go [hunk 11]
```diff
@@ -94,127 +94,43 @@ func convertImagesToWebsocketScanCommand(registry *registryScan, sessionObj *uti
...
-	registryName, err := actionHandler.parseRegistryNameArg(sessionObj)
+func (actionHandler *ActionHandler) testRegistryConnectivity(sessionObj *utils.SessionObj) error {
+	registryScan, err := actionHandler.loadRegistryScan(sessionObj)
 	if err != nil {
-		return fmt.Errorf("parseRegistryNameArg failed with err %v", err)
-	}
-	auth, err := actionHandler.getRegistryAuth(registryName)
-	if err != nil {
-		return fmt.Errorf("get registry auth failed with err %v", err)
+		glog.Errorf("in testRegistryConnectivity: loadRegistryScan failed with error: %v", err.Error())
```
Comment (Bezbran) [inline mainhandler/vulnscan.go:125]: should we update the job report details

**Review by Bezbran — APPROVED** [review]

File: mainhandler/imageregistryhandler.go (hunk 8 above)
Comment (Daniel-GrunbergerCA) [reply mainhandler/imageregistryhandler.go:432]: it won't work, token has single dots before

File: mainhandler/imageregistryhandler.go (hunk 9 above)
Comment (Daniel-GrunbergerCA) [reply mainhandler/imageregistryhandler.go:459]: If we get an error, we won't have nextPage. We do set details based on different errors for test connectivity

File: mainhandler/imageregistryhandler.go (hunk 10 above)
Comment (Daniel-GrunbergerCA) [reply mainhandler/imageregistryhandler.go:658]: This flow will only be called if we checked the command and it is the old one, meaning we need to take all info from cluster

File: mainhandler/vulnscan.go (hunk 11 above)
Comment (Daniel-GrunbergerCA) [reply mainhandler/vulnscan.go:125]: This is error in command so I didn't use the new statuses on it at all. The error will be sent as error in the job report. Should I update details which the step failed as well?

Comment (Bezbran) [reply mainhandler/vulnscan.go:125]: Yes

File: mainhandler/imageregistryhandler.go (hunk 9 above)
Comment (Bezbran) [reply mainhandler/imageregistryhandler.go:459]: Details based on steps/stages not on errors

**Review by Bezbran — APPROVED** [review]

**Pull request merged**